In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("online_retail_II.csv")

In [3]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [4]:
df.shape

(1067371, 8)

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  str    
 1   StockCode    1067371 non-null  str    
 2   Description  1062989 non-null  str    
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  str    
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 136.6 MB


In [6]:
df["Customer ID"].isna().sum()

np.int64(243007)

In [7]:
missing = df["Customer ID"].isna().sum()
total = len(df)

missing / total * 100

np.float64(22.766872999172733)

# Data Understanding

## Findings

- The dataset contains **1,067,371 transactions** and **8 columns**.
- Approximately **22.8%** of transactions are missing a Customer ID.
- Since our objective is customer segmentation, these records will be removed because they cannot be linked to an individual customer.

In [8]:
rows_before = len(df)
rows_before

1067371

In [9]:
df = df.dropna(subset=["Customer ID"])

In [10]:
len(df)

824364

### Data Cleaning Insight

Rows with missing **Customer ID** values were removed because they cannot be associated with a specific customer. Since the objective of this project is customer segmentation, these records cannot contribute to customer-level analysis.

- Rows before cleaning: 1,067,371
- Rows after cleaning: 824,364
- Rows removed: 243,007 (22.8%)

### Data Cleaning Step 1: Missing Customer IDs

The dataset originally contained **1,067,371** transactions. An analysis of missing values showed that **243,007 transactions (22.8%)** did not have a Customer ID.

Because the objective of this project is customer segmentation, these transactions cannot be associated with a specific customer. Therefore, they were removed from the dataset.

After cleaning, a verification check confirmed that there are **0 missing Customer ID values** remaining.

In [11]:
df["Customer ID"].isna().sum()

np.int64(0)

In [12]:
(df["Quantity"] < 0).sum()

np.int64(18744)

In [13]:
df[df["Quantity"] < 0].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
183,C489449,21871,SAVE THE PLANET MUG,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,2009-12-01 10:33:00,0.85,16321.0,Australia
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,2009-12-01 10:44:00,4.25,17592.0,United Kingdom


### Data Cleaning Step 2: Negative Quantities

An investigation identified **18,744** transactions with negative quantities.

These transactions have invoice numbers beginning with **"C"**, indicating they are credit notes or cancellations. Since the objective of this project is to analyze completed customer purchases for segmentation, these transactions will be excluded from the analysis.


In [14]:
df = df[df["Quantity"] > 0]

In [15]:
(df["Quantity"] < 0).sum()

np.int64(0)

In [16]:
df.shape

(805620, 8)

### Data Cleaning Summary

The original dataset contained **1,067,371** transactions.

The following cleaning steps were performed:

1. Removed **243,007** transactions with missing Customer IDs because they could not be linked to individual customers.
2. Removed **18,744** transactions with negative quantities, which were identified as cancellations or credit notes.

The cleaned dataset contains **805,620** completed purchase transactions and is ready for analysis.

In [17]:
(df["Price"] <= 0).sum()

np.int64(71)

In [18]:
df[df["Price"] <= 0].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
4674,489825,22076,6 RIBBONS EMPIRE,12,2009-12-02 13:34:00,0.0,16126.0,United Kingdom
6781,489998,48185,DOOR MAT FAIRY CAKE,2,2009-12-03 11:19:00,0.0,15658.0,United Kingdom
16107,490727,M,Manual,1,2009-12-07 16:38:00,0.0,17231.0,United Kingdom
18738,490961,22065,CHRISTMAS PUDDING TRINKET POT,1,2009-12-08 15:25:00,0.0,14108.0,United Kingdom
18739,490961,22142,CHRISTMAS CRAFT WHITE FAIRY,12,2009-12-08 15:25:00,0.0,14108.0,United Kingdom
32916,492079,85042,ANTIQUE LILY FAIRY LIGHTS,8,2009-12-15 13:49:00,0.0,15070.0,United Kingdom
40101,492760,21143,ANTIQUE GLASS HEART DECORATION,12,2009-12-18 14:22:00,0.0,18071.0,United Kingdom
47126,493761,79320,FLAMINGO LIGHTS,24,2010-01-06 14:54:00,0.0,14258.0,United Kingdom
48342,493899,22355,"CHARLOTTE BAG , SUKI DESIGN",10,2010-01-08 10:43:00,0.0,12417.0,Belgium
57619,494607,21533,RETRO SPOT LARGE MILK JUG,12,2010-01-15 12:43:00,0.0,16858.0,United Kingdom


### Data Cleaning Step 3: Zero-Priced Transactions

A total of **71** completed transactions had a unit price of zero.

These transactions likely represent promotional items, manual adjustments, or complimentary products. Since the objective of this project is to analyze customer spending behavior, these records were excluded because they do not contribute to monetary value.

Removing these records has a negligible impact on the dataset due to their very small proportion.

In [19]:
df = df[df["Price"] > 0]

In [20]:
(df["Price"] <= 0).sum()

np.int64(0)

In [21]:
df["Customer ID"].nunique()

5878

In [22]:
df["Country"].value_counts().head(10)

Country
United Kingdom    725250
Germany            16694
EIRE               15743
France             13812
Netherlands         5088
Spain               3719
Belgium             3068
Switzerland         3011
Portugal            2446
Australia           1812
Name: count, dtype: int64

In [23]:
uk_transactions = df[df["Country"] == "United Kingdom"].shape[0]
total_transactions = len(df)

uk_transactions / total_transactions * 100

90.03176715507064

In [24]:
df["Revenue"] = df["Quantity"] * df["Price"]

In [25]:
df["Revenue"].sum()

np.float64(17743429.178000003)

### Exploratory Data Analysis: Total Revenue

After removing incomplete, canceled, and zero-priced transactions, the cleaned dataset generated a total revenue of **£17,743,429.18**.

This represents the value of completed purchases and will be used throughout the remainder of the analysis, particularly when calculating customer monetary value (RFM).

In [26]:
top_customers = (
    df.groupby("Customer ID")["Revenue"]
      .sum()
      .sort_values(ascending=False)
      .head(10)
)

top_customers

Customer ID
18102.0    608821.65
14646.0    528602.52
14156.0    313946.37
14911.0    295972.63
17450.0    246973.09
13694.0    196482.81
17511.0    175603.55
16446.0    168472.50
16684.0    147142.77
12415.0    144458.37
Name: Revenue, dtype: float64

### Exploratory Data Analysis: Top Revenue-Generating Customers

The analysis identified the ten highest-spending customers based on total revenue.

The highest-value customer generated over **£608,000** in revenue, while several others contributed more than **£100,000** each.

This indicates that customer spending varies substantially, suggesting that targeted marketing strategies based on customer value could be beneficial.

# Next Step: RFM Analysis

The dataset has been cleaned and explored. The next phase will transform transaction-level data into customer-level metrics by calculating Recency, Frequency, and Monetary (RFM) values for each customer.

In [27]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom,83.4
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom,81.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom,100.8
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom,30.0


In [28]:
import pandas as pd

# Load the dataset
df = pd.read_csv("online_retail_II.csv")

# Display the first five rows
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [29]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  str    
 1   StockCode    1067371 non-null  str    
 2   Description  1062989 non-null  str    
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  str    
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 136.6 MB


In [30]:
df.columns

Index(['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'Price', 'Customer ID', 'Country'],
      dtype='str')

In [31]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype  
---  ------       --------------    -----  
 0   Invoice      1067371 non-null  str    
 1   StockCode    1067371 non-null  str    
 2   Description  1062989 non-null  str    
 3   Quantity     1067371 non-null  int64  
 4   InvoiceDate  1067371 non-null  str    
 5   Price        1067371 non-null  float64
 6   Customer ID  824364 non-null   float64
 7   Country      1067371 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 136.6 MB


In [32]:
import os

for file in os.listdir():
    print(file)

.ipynb_checkpoints
cleaned_online_retail.csv
CLEANING DATA.ipynb
customer_rfm.csv
online_retail_II.csv


In [33]:
df = df.dropna(subset=["Customer ID"]).copy()

In [34]:
df["Customer ID"].isna().sum()

np.int64(0)

In [35]:
import os

In [36]:
for file in os.listdir():
    print(file)

.ipynb_checkpoints
cleaned_online_retail.csv
CLEANING DATA.ipynb
customer_rfm.csv
online_retail_II.csv


In [37]:
df.info()

<class 'pandas.DataFrame'>
Index: 824364 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Invoice      824364 non-null  str    
 1   StockCode    824364 non-null  str    
 2   Description  824364 non-null  str    
 3   Quantity     824364 non-null  int64  
 4   InvoiceDate  824364 non-null  str    
 5   Price        824364 non-null  float64
 6   Customer ID  824364 non-null  float64
 7   Country      824364 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 112.2 MB


In [38]:
df = df.dropna(subset=["Customer ID"])

In [39]:
import pandas as pd

df = pd.read_csv("online_retail_II.csv")

df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [40]:
print("Original shape:", df.shape)

Original shape: (1067371, 8)


In [41]:
df = df.dropna(subset=["Customer ID"]).copy()

print("Shape after removing missing Customer IDs:", df.shape)
print("Missing Customer IDs:", df["Customer ID"].isna().sum())

Shape after removing missing Customer IDs: (824364, 8)
Missing Customer IDs: 0


In [42]:
df["Customer ID"] = df["Customer ID"].astype(int)

In [43]:
df["InvoiceDate"] = pd.to_datetime(df["InvoiceDate"], errors="coerce")

In [44]:
df["InvoiceDate"].dtype

dtype('<M8[us]')

In [45]:
df = df[~df["Invoice"].astype(str).str.startswith("C")].copy()

print("Shape after removing cancelled transactions:", df.shape)

Shape after removing cancelled transactions: (805620, 8)


In [46]:
print(df["Invoice"].astype(str).str.startswith("C").sum())

0


In [47]:
df.to_csv("cleaned_online_retail.csv", index=False)

In [48]:
import os
print(os.path.exists("cleaned_online_retail.csv"))

True


In [49]:
# Calculate total revenue for each customer
customer_revenue = (
    df.groupby("Customer ID")["Revenue"]
      .sum()
      .sort_values(ascending=False)
)

# Display the top 10 customers
customer_revenue.head(10)

KeyError: 'Column not found: Revenue'

In [ ]:
df.columns

In [ ]:
# Create a Revenue column
df["Revenue"] = df["Quantity"] * df["Price"]

In [ ]:
df.columns

In [ ]:
# Calculate total revenue for each customer
customer_revenue = (
    df.groupby("Customer ID")["Revenue"]
      .sum()
      .sort_values(ascending=False)
)

customer_revenue.head(10)

In [ ]:
import matplotlib.pyplot as plt

# Select the top 10 customers
top10 = customer_revenue.head(10)

# Create the bar chart
plt.figure(figsize=(10, 6))
plt.bar(top10.index.astype(str), top10.values)

plt.title("Top 10 Revenue-Generating Customers")
plt.xlabel("Customer ID")
plt.ylabel("Total Revenue (£)")
plt.xticks(rotation=45)

plt.tight_layout()
plt.show()

Exploratory Data Analysis: Top Revenue-Generating Customers

Total revenue was calculated for each customer by summing the value of all completed purchases. The analysis identified the ten highest-spending customers, with Customer 18102 generating approximately £608,822 in revenue, followed by Customer 14646 with approximately £528,603. Several other customers also generated over £140,000 in total revenue. These results indicate that customer spending is highly uneven, with a relatively small number of customers contributing a substantial proportion of total sales. This finding supports the use of customer segmentation techniques, such as RFM analysis, to identify high-value customers and develop targeted marketing and retention strategies.

In [ ]:
# Create a snapshot date (one day after the most recent transaction)
snapshot_date = df["InvoiceDate"].max() + pd.Timedelta(days=1)

print(snapshot_date)

In [ ]:
# Create the RFM table
rfm = df.groupby("Customer ID").agg({
    "InvoiceDate": lambda x: (snapshot_date - x.max()).days,
    "Invoice": "nunique",
    "Revenue": "sum"
})

# Rename the columns
rfm.columns = ["Recency", "Frequency", "Monetary"]

# Display the first five rows
rfm.head()

In [ ]:
rfm.head()

In [ ]:
# Summary statistics
rfm.describe()

In [ ]:
import matplotlib.pyplot as plt

rfm.hist(figsize=(12, 4), bins=30)

plt.tight_layout()
plt.show()

In [ ]:
# Recency score (lower recency = better)
rfm["R_Score"] = pd.qcut(
    rfm["Recency"],
    q=5,
    labels=[5, 4, 3, 2, 1]
).astype(int)

# Frequency score (higher frequency = better)
rfm["F_Score"] = pd.qcut(
    rfm["Frequency"].rank(method="first"),
    q=5,
    labels=[1, 2, 3, 4, 5]
).astype(int)

# Monetary score (higher spending = better)
rfm["M_Score"] = pd.qcut(
    rfm["Monetary"],
    q=5,
    labels=[1, 2, 3, 4, 5]
).astype(int)

# Combine into one RFM score
rfm["RFM_Score"] = (
    rfm["R_Score"].astype(str)
    + rfm["F_Score"].astype(str)
    + rfm["M_Score"].astype(str)
)

rfm.head()

In [ ]:
# Function to assign customer segments
def customer_segment(row):
    if row["R_Score"] >= 4 and row["F_Score"] >= 4 and row["M_Score"] >= 4:
        return "Champions"
    elif row["R_Score"] >= 3 and row["F_Score"] >= 4:
        return "Loyal Customers"
    elif row["R_Score"] >= 4 and row["F_Score"] >= 2:
        return "Potential Loyalists"
    elif row["R_Score"] <= 2 and row["F_Score"] >= 3:
        return "At Risk"
    else:
        return "Hibernating"

# Apply the function
rfm["Segment"] = rfm.apply(customer_segment, axis=1)

In [ ]:
rfm["Segment"].value_counts()

In [ ]:
def customer_segment(row):
    if row["R_Score"] >= 4 and row["F_Score"] >= 4 and row["M_Score"] >= 4:
        return "High-Value Customers"
    elif row["R_Score"] >= 3 and row["F_Score"] >= 4:
        return "Loyal Customers"
    elif row["R_Score"] >= 4 and row["F_Score"] >= 2:
        return "Emerging Customers"
    elif row["R_Score"] <= 2 and row["F_Score"] >= 3:
        return "At Risk"
    else:
        return "Inactive Customers"

rfm["Segment"] = rfm.apply(customer_segment, axis=1)

In [ ]:
rfm["Segment"].value_counts()

In [ ]:
segment_order = [
    "High-Value Customers",
    "Loyal Customers",
    "Emerging Customers",
    "At Risk",
    "Inactive Customers"
]

segment_counts = (
    rfm["Segment"]
    .value_counts()
    .reindex(segment_order)
)

print(segment_counts)

In [ ]:
import matplotlib.pyplot as plt

segment_order = [
    "High-Value Customers",
    "Loyal Customers",
    "Emerging Customers",
    "At Risk",
    "Inactive Customers"
]

segment_counts = (
    rfm["Segment"]
    .value_counts()
    .reindex(segment_order)
)

plt.figure(figsize=(8,5))
plt.bar(segment_counts.index, segment_counts.values)

plt.title("Customer Segments")
plt.xlabel("Customer Segment")
plt.ylabel("Number of Customers")

plt.xticks(rotation=20)

plt.tight_layout()
plt.show()

Customer Segmentation Analysis

The RFM analysis grouped the 5,881 customers into five distinct customer segments based on their purchasing behaviour. As shown in the chart, Inactive Customers represent the largest segment, with 2,345 customers. This indicates that a significant proportion of customers have not purchased recently or have low purchasing activity, suggesting opportunities for re-engagement campaigns.

The second largest segment is High-Value Customers with 1,300 customers. These customers purchase frequently, have made recent purchases, and generate the highest levels of revenue. Retaining this group should be a priority through loyalty programmes, personalised offers, and exclusive promotions.

The At Risk segment contains 824 customers. These customers have previously demonstrated valuable purchasing behaviour but have not purchased recently, making them suitable targets for win-back campaigns.

The Emerging Customers segment consists of 714 customers who have recently engaged with the business and show potential to become loyal customers if nurtured through targeted marketing initiatives.

Finally, Loyal Customers account for 698 customers. Although this is the smallest segment, these customers purchase consistently and represent a stable source of revenue. Maintaining strong relationships with this group through continued engagement and rewards can help encourage long-term customer retention.

Overall, the customer segmentation demonstrates that different groups exhibit different purchasing behaviours. This enables the business to develop tailored marketing strategies for each segment, improving customer retention, increasing customer lifetime value, and allocating marketing resources more effectively.

Key Business Insights

1. A large proportion of customers are inactive.

2,345 of the 5,881 customers were classified as Inactive Customers.
This suggests an opportunity to improve customer retention through re-engagement campaigns.

2. High-value customers drive significant business value.

1,300 customers were classified as High-Value Customers.
Retaining these customers should be a priority through loyalty programs and personalized offers.

3. At-Risk customers present a recovery opportunity.

824 customers previously demonstrated valuable purchasing behaviour but have not purchased recently.
Targeted promotions or reminder emails could encourage them to return.

4. Emerging Customers have growth potential.

714 customers have started engaging with the business and could become loyal customers with effective marketing.

5. Loyal Customers provide consistent revenue.

Although smaller in number, these customers make repeat purchases and contribute to long-term business stability.

In [ ]:
df.to_csv("cleaned_online_retail.csv", index=False)

In [ ]:
rfm.to_csv("customer_rfm.csv")

In [ ]:
# Save the cleaned transaction data
df.to_csv("cleaned_online_retail.csv", index=False)

In [ ]:
# Save the customer-level RFM table
rfm.to_csv("customer_rfm.csv")

In [ ]:
import os

print(os.listdir())

In [ ]:
rfm.to_csv("customer_rfm.csv")

In [50]:
rfm.head()

NameError: name 'rfm' is not defined

In [51]:
df["Revenue"] = df["Quantity"] * df["Price"]

In [52]:
snapshot_date = df["InvoiceDate"].max() + pd.Timedelta(days=1)
print(snapshot_date)

2011-12-10 12:50:00


In [53]:
rfm = df.groupby("Customer ID").agg({
    "InvoiceDate": lambda x: (snapshot_date - x.max()).days,
    "Invoice": "nunique",
    "Revenue": "sum"
})

rfm.columns = ["Recency", "Frequency", "Monetary"]

rfm.head()

,Recency,Frequency,Monetary
Customer ID,,,
12346,326,12,77556.46
12347,2,8,5633.32
12348,75,5,2019.40
12349,19,4,4428.69
12350,310,1,334.40


In [54]:
rfm.to_csv("customer_rfm.csv")

In [55]:
import os
print(os.path.exists("customer_rfm.csv"))

True


In [56]:
rfm.columns

Index(['Recency', 'Frequency', 'Monetary'], dtype='str')

In [57]:
R_Score
F_Score
M_Score

NameError: name 'R_Score' is not defined

In [58]:
rfm.columns

Index(['Recency', 'Frequency', 'Monetary'], dtype='str')

In [59]:
# Recency score (lower recency = better)
rfm["R_Score"] = pd.qcut(
    rfm["Recency"],
    q=5,
    labels=[5, 4, 3, 2, 1]
).astype(int)

# Frequency score (higher frequency = better)
rfm["F_Score"] = pd.qcut(
    rfm["Frequency"].rank(method="first"),
    q=5,
    labels=[1, 2, 3, 4, 5]
).astype(int)

# Monetary score (higher spending = better)
rfm["M_Score"] = pd.qcut(
    rfm["Monetary"],
    q=5,
    labels=[1, 2, 3, 4, 5]
).astype(int)

rfm["RFM_Score"] = (
    rfm["R_Score"].astype(str)
    + rfm["F_Score"].astype(str)
    + rfm["M_Score"].astype(str)
)

In [60]:
def customer_segment(row):
    if row["R_Score"] >= 4 and row["F_Score"] >= 4 and row["M_Score"] >= 4:
        return "High-Value Customers"
    elif row["R_Score"] >= 3 and row["F_Score"] >= 4:
        return "Loyal Customers"
    elif row["R_Score"] >= 4 and row["F_Score"] >= 2:
        return "Emerging Customers"
    elif row["R_Score"] <= 2 and row["F_Score"] >= 3:
        return "At Risk"
    else:
        return "Inactive Customers"

rfm["Segment"] = rfm.apply(customer_segment, axis=1)

In [61]:
rfm.columns

Index(['Recency', 'Frequency', 'Monetary', 'R_Score', 'F_Score', 'M_Score',
       'RFM_Score', 'Segment'],
      dtype='str')

In [62]:
rfm["Segment"].value_counts()

Segment
Inactive Customers      2345
High-Value Customers    1300
At Risk                  824
Emerging Customers       714
Loyal Customers          698
Name: count, dtype: int64

In [63]:
rfm.to_csv("customer_rfm_dashboard.csv")

In [64]:
import os
print(os.path.exists("customer_rfm_dashboard.csv"))

True
